File: step_01_basics/1.1_logic_building/P010_pattern_10_half_diamond_star/Solution.java
/**
 * Problem: Pattern 10 - Half Diamond Star
 * Difficulty: EASY | XP: 10
 *
 * Print a left-aligned half diamond: stars increase from 1 to N,
 * then decrease from N-1 back to 1. Total (2*N - 1) rows.
 *
 * @author DSA_Nova
 */
public class Solution {

    // ============================================================
    // APPROACH 1: BRUTE FORCE -- Two Separate Loops
    // Time: O(N^2)  |  Space: O(1)
    // First loop: increasing 1..N. Second loop: decreasing N-1..1.
    // ============================================================
    static class BruteForce {
        public static void printPattern(int n) {
            // Increasing half: 1 to N stars
            for (int i = 1; i <= n; i++) {
                for (int j = 0; j < i; j++) {
                    if (j > 0) System.out.print(" ");
                    System.out.print("*");
                }
                System.out.println();
            }
            // Decreasing half: N-1 down to 1 star
            for (int i = n - 1; i >= 1; i--) {
                for (int j = 0; j < i; j++) {
                    if (j > 0) System.out.print(" ");
                    System.out.print("*");
                }
                System.out.println();
            }
        }
    }

    // ============================================================
    // APPROACH 2: OPTIMAL -- Single Loop with abs() Formula
    // Time: O(N^2)  |  Space: O(N)
    // stars = N - |N - 1 - i| for i in [0, 2N-2].
    // ============================================================
    static class Optimal {
        public static void printPattern(int n) {
            int totalRows = 2 * n - 1;
            for (int i = 0; i < totalRows; i++) {
                int stars = n - Math.abs(n - 1 - i);
                StringBuilder row = new StringBuilder();
                for (int j = 0; j < stars; j++) {
                    if (j > 0) row.append(" ");
                    row.append("*");
                }
                System.out.println(row);
            }
        }
    }

    // ============================================================
    // APPROACH 3: BEST -- Full Grid String, Single Print
    // Time: O(N^2)  |  Space: O(N^2)
    // Build entire half diamond as one string, single I/O call.
    // ============================================================
    static class Best {
        public static void printPattern(int n) {
            StringBuilder grid = new StringBuilder();
            int totalRows = 2 * n - 1;
            for (int i = 0; i < totalRows; i++) {
                if (i > 0) grid.append("\n");
                int stars = n - Math.abs(n - 1 - i);
                for (int j = 0; j < stars; j++) {
                    if (j > 0) grid.append(" ");
                    grid.append("*");
                }
            }
            System.out.println(grid);
        }
    }

    public static void main(String[] args) {
        int n = 5;
        System.out.println("=== Pattern 10 - Half Diamond Star ===");

        System.out.println("--- Brute Force (N=5) ---");
        BruteForce.printPattern(n);

        System.out.println("--- Optimal (N=5) ---");
        Optimal.printPattern(n);

        System.out.println("--- Best (N=5) ---");
        Best.printPattern(n);

        // Edge cases
        System.out.println("--- N=1 ---");
        Best.printPattern(1);

        System.out.println("--- N=3 ---");
        Best.printPattern(3);
    }
}

File: step_01_basics/1.1_logic_building/P010_pattern_10_half_diamond_star/solution.py
"""
Problem: Pattern 10 - Half Diamond Star
Difficulty: EASY | XP: 10

Print a left-aligned half diamond: stars increase from 1 to N,
then decrease from N-1 back to 1. Total (2*N - 1) rows.
"""


# ============================================================
# APPROACH 1: BRUTE FORCE -- Two Separate Loops
# Time: O(N^2)  |  Space: O(1)
# First loop: increasing 1..N. Second loop: decreasing N-1..1.
# ============================================================
def brute_force(n: int) -> None:
    # Increasing half
    for i in range(1, n + 1):
        for j in range(i):
            if j > 0:
                print(" ", end="")
            print("*", end="")
        print()
    # Decreasing half
    for i in range(n - 1, 0, -1):
        for j in range(i):
            if j > 0:
                print(" ", end="")
            print("*", end="")
        print()


# ============================================================
# APPROACH 2: OPTIMAL -- Single Loop with abs() Formula
# Time: O(N^2)  |  Space: O(N)
# stars = N - |N - 1 - i| for i in [0, 2N-2].
# ============================================================
def optimal(n: int) -> None:
    total_rows = 2 * n - 1
    for i in range(total_rows):
        stars = n - abs(n - 1 - i)
        print(" ".join(["*"] * stars))


# ============================================================
# APPROACH 3: BEST -- Full Grid String, Single Print
# Time: O(N^2)  |  Space: O(N^2)
# Build entire half diamond as one string, single I/O call.
# ============================================================
def best(n: int) -> None:
    total_rows = 2 * n - 1
    rows = []
    for i in range(total_rows):
        stars = n - abs(n - 1 - i)
        rows.append(" ".join(["*"] * stars))
    print("\n".join(rows))


if __name__ == "__main__":
    N = 5
    print("=== Pattern 10 - Half Diamond Star ===")

    print("--- Brute Force (N=5) ---")
    brute_force(N)

    print("--- Optimal (N=5) ---")
    optimal(N)

    print("--- Best (N=5) ---")
    best(N)

    # Edge cases
    print("--- N=1 ---")
    best(1)

    print("--- N=3 ---")
    best(3)
